sqlite.go

```diff
@@ -0,0 +1,530 @@
+	return &Error{
+		Code:  sqliteh.SQLITE_MISUSE,
+		Loc:   "Bind",
+		Query: s.query,
+		Msg:   fmt.Sprintf("Bind:%v: unknown value type %T", debugName, arg.Value),
+	}
+}
+
+func (s *stmt) bindBasic(debugName interface{}, ordinal int, v interface{}) (found bool, err error) {
+	defer func() {
+		if err != nil {
+			err = s.reserr(fmt.Sprintf("Bind:%v:%T", debugName, v), err)
+		}
+	}()
+	switch v := v.(type) {
+	case nil:
+		return true, s.stmt.BindNull(ordinal)
+	case string:
+		return true, s.stmt.BindText64(ordinal, v)
+	case int:
+		return true, s.stmt.BindInt64(ordinal, int64(v))
+	case int64:
+		return true, s.stmt.BindInt64(ordinal, v)
+	case float64:
+		return true, s.stmt.BindDouble(ordinal, v)
+	case []byte:
+		if len(v) == 0 {
+			return true, s.stmt.BindZeroBlob64(ordinal, 0)
+		} else {
+			return true, s.stmt.BindBlob64(ordinal, v)
+		}
+	case time.Time:
+		// Shortest of:
+		//	YYYY-MM-DD HH:MM
+		// 	YYYY-MM-DD HH:MM:SS
+		//	YYYY-MM-DD HH:MM:SS.SSS
+		str := v.Format(TimeFormat)
+		str = strings.TrimSuffix(str, "-0000")
+		str = strings.TrimSuffix(str, ".000")
+		str = strings.TrimSuffix(str, ":00")
+		return true, s.stmt.BindText64(ordinal, str)
+	default:
+		return false, nil
+	}
+}
+
+type rows struct {
+	stmt   *stmt
+	closed bool
+
+	colNames []string // filled on call to Columns
+
+	// Filled on first call to Next.
+	colTypes     []sqliteh.ColumnType
+	colDeclTypes []string
+}
+
+func (r *rows) Columns() []string {
+	if r.closed {
+		panic("Columns called after Rows was closed")
+	}
+	if r.colNames == nil {
+		r.colNames = make([]string, r.stmt.stmt.ColumnCount())
+		for i := range r.colNames {
+			r.colNames[i] = r.stmt.stmt.ColumnName(i)
+		}
+	}
+	return append([]string{}, r.colNames...)
+}
+
+func (r *rows) Close() error {
+	if r.closed {
+		return errors.New("sqlite rows result already closed")
+	}
+	r.closed = true
+	if err := r.stmt.stmt.Reset(); err != nil {
+		return r.stmt.reserr("Rows.Close(Reset)", err)
+	}
+	return r.stmt.reserr("Rows.Close(ClearBindings)", r.stmt.stmt.ClearBindings())
+}
+
+func (r *rows) Next(dest []driver.Value) error {
+	if r.closed {
+		return errors.New("sqlite rows result already closed")
+	}
+	hasRow, err := r.stmt.stmt.Step()
+	if err != nil {
+		return err
+	}
+	if !hasRow {
+		return io.EOF
+	}
+
+	if r.colTypes == nil {
+		colCount := r.stmt.stmt.ColumnCount()
+		r.colTypes = make([]sqliteh.ColumnType, colCount)
+		r.colDeclTypes = make([]string, colCount)
+
+		for i := range r.colTypes {
+			r.colTypes[i] = r.stmt.stmt.ColumnType(i)
+			r.colDeclTypes[i] = r.stmt.stmt.ColumnDeclType(i)
+		}
+	}
+
+	for i := range dest {
+		switch r.colDeclTypes[i] {
+		case "DATETIME", "DATE":
+			switch r.colTypes[i] {
+			case sqliteh.SQLITE_INTEGER:
+				v := r.stmt.stmt.ColumnInt64(i)
+				dest[i] = time.Unix(v, 0)
+			case sqliteh.SQLITE_FLOAT:
+				dest[i] = r.stmt.stmt.ColumnDouble(i)
+			case sqliteh.SQLITE_TEXT:
+				v := r.stmt.stmt.ColumnText(i)
+				t, err := time.Parse(TimeFormat, v)
```

Comment:
time.Parse will handle truncated inputs correctly? That's neat.

Reply:
So I'm sure I tested this. But apparently it doesn't work? Oops.